### ISSUE TITLE
AccountInfo::assign is unsound

### ISSUE THREAD

**Issue by brson** [solana-labs/solana#28607]
#### Problem

Here's the definition:

```rust
    pub fn assign(&self, new_owner: &Pubkey) {
        // Set the non-mut owner field
        unsafe {
            std::ptr::write_volatile(
                self.owner as *const Pubkey as *mut [u8; 32],
                new_owner.to_bytes(),
            );
        }
    }
```

This is not allowed in Rust and causes undefined behavior. It's not clear what `write_volatile` is doing here, but it's probably not what is intended.

#### Proposed Solution

I don't see any way to do this without changing the definition of `AccountInfo`.

**Comment by jackcmay** [comment]
Right, we are not able to change `AccountInfo`, but is being replaced as part of the ABIv2 effort

https://github.com/solana-labs/solana/issues/27384

**Comment by steveluscher** [comment]
Can this be closed as sufficiently covered by #27384, @Lichtso?

**Comment by Lichtso** [comment]
yes
